client/extensions/woocommerce/app/settings/email/mailchimp/setup-mailchimp.js

```diff
@@ -0,0 +1,324 @@
+	'store_currency_code', 'admin_email' ];
+
+const campaignDefaultsRequiredFields = [ 'campaign_from_name', 'campaign_from_email', 'campaign_subject',
+		'campaign_language', 'campaign_permission_reminder' ];
+
+class MailChimpSetup extends React.Component {
+
+	constructor( props ) {
+		super( props );
+		// make this react to the real phase the execution is.
+		this.state = {
+			step: LOG_INTO_MAILCHIMP_STEP,
+			settings: this.prepareDefaultValues( this.props ),
+			settings_values_missing: false,
+			api_key_input: this.props.settings.mailchimp_api_key || '',
+		};
+	}
+
+	componentWillReceiveProps( nextProps ) {
+		// No state changes while doing request.
+		if ( nextProps.isBusy ) {
+			return;
+		}
+
+		const { active_tab } = nextProps.settings;
+		if ( steps[ this.state.step ].nextStep === active_tab ) {
+			const settings = this.prepareDefaultValues( nextProps );
+			this.setState( { step: active_tab, settings } );
+			if ( active_tab === STORE_SYNC ) {
+				nextProps.onClose();
+			}
+		}
+
+		// Update settings if we have received lists
+		if ( nextProps.settings.mailchimp_lists && ! this.props.settings.mailchimp_lists ) {
+			const settings = this.prepareDefaultValues( nextProps );
+			this.setState( { step: active_tab, settings } );
+		}
+	}
+
+	prepareDefaultValues( nextProps ) {
+		const { settings } = nextProps;
+		const newSettings = Object.assign( {}, settings );
+		newSettings.campaign_from_name = settings.campaign_from_name || settings.store_name || '';
+		newSettings.campaign_from_email = settings.campaign_from_email || settings.admin_email || '';
+		newSettings.campaign_subject = settings.campaign_subject || settings.store_name || '';
+		newSettings.store_locale = settings.store_locale || 'en';
+		newSettings.campaign_language = settings.campaign_language || settings.store_locale;
+		newSettings.campaign_permission_reminder = settings.campaign_permission_reminder ||
+			'You were subscribed to the newsletter from ' + settings.store_name;
+		newSettings.admin_email = settings.admin_email || nextProps.currentUserEmail || '';
+		newSettings.store_timezone = settings.store_timezone || nextProps.timezone || 'America/New_York';
+		newSettings.mailchimp_lists = settings.mailchimp_lists;
+		newSettings.mailchimp_list = settings.mailchimp_list;
+		return newSettings;
+	}
+
+	getStoreSettings = () => {
+		// clear this and pass only what is required.
+		const { address, currency } = this.props;
+
+		const settings = pick( this.state.settings, storeSettingsRequiredFields );
+		settings.store_city = address.city;
+		settings.store_street = address.street + ' ' + address.street2;
+		settings.store_state = address.state;
+		settings.store_country = address.country;
+		settings.store_postal_code = address.postcode;
+		settings.store_currency_code = currency;
+		return settings;
+	}
+
+	getCampaingDefaultsSettings = () => {
+		return pick( this.state.settings, campaignDefaultsRequiredFields );
+	}
+
+	hasEmptyValues = ( data ) => {
```

Comment:
Looks like this is only used once, perhaps just inline this logic on line 136?

Reply:
Added this check to one more place so it makes sense now to keep it.